### PR TITLE
KAFKA-9875: Make integration tests more resilient

### PR DESCRIPTION
**Pull request by vvcephei**
The ticket is for a flaky test that failed to clean up topics _after_ the test, which
isn't strictly necessary for test success.

* alter the "clean up after test" method to never throw an exception
  (after verifying it's always the last invocation inside a finally block,
  so it won't break any test semantics)
* consolidated the naming of all integration tests' app ids, topics, etc., by introducing 
  a new test utility to generate safe, unique, descriptive names.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java [hunk 1]
```diff
@@ -101,8 +102,8 @@ public void before() throws Exception {
         builder = new StreamsBuilder();
         createTopics();
         streamsConfiguration = new Properties();
-        final String applicationId = "global-thread-shutdown-test" + testName.getMethodName();
-        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        final String safeTestName = safeUniqueTestName(getClass(), testName);
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "app-" + safeTestName);
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalThreadShutDownOrderTest.java:106]: I've standardized all the usages to be just "app", followed by the generated name, since the generated name contains the same information that we previously hand-wrote into the prefix or suffix. All we really need to do is ensure that the app id won't collide with a group name that we might use in a verification consumer, for example. For that reason, I've never used the generated name "plain", but always scoped it to the usage (app id, group id, input topic, etc.).

It's not super important to apply these ideas universally, but I felt it would make it easier to write more tests like it in the future if I just made a full pass on all the tests to make them all look the same.

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java [hunk 2]
```diff
@@ -88,16 +89,17 @@
...
-        outputTopicName = "output-topic-" + name.getMethodName();
-        stateStoreName = "lagfetch-test-store" + name.getMethodName();
+        final String safeTestName = safeUniqueTestName(getClass(), testName);
+        inputTopicName = "input-topic-" + safeTestName;
+        outputTopicName = "output-topic-" + safeTestName;
+        stateStoreName = "lagfetch-test-store" + safeTestName;
 
         streamsConfiguration = new Properties();
-        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "lag-fetch-" + name.getMethodName());
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "lag-fetch-" + safeTestName);
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]:
```suggestion
        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "app-" + safeTestName);
```

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 3]
```diff
@@ -145,12 +162,12 @@ public static void cleanStateBeforeTest(final EmbeddedKafkaCluster cluster, fina
...
-        driver.cleanUp();
+    public static void quietlyCleanStateAfterTest(final EmbeddedKafkaCluster cluster, final KafkaStreams driver) {
         try {
+            driver.cleanUp();
             cluster.deleteAllTopicsAndWait(DEFAULT_TIMEOUT);
-        } catch (final InterruptedException e) {
-            throw new RuntimeException(e);
+        } catch (final RuntimeException | InterruptedException e) {
+            LOG.warn("Ignoring failure to clean test state", e);
         }
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:171]: This is really the fix for KAFKA-9875. The other change just hopefully reduces the probability that ignoring the exceptions could cause subsequent failures (e.g., if the topics don't get deleted before the next test, at least the next one will have different topic names).

I've verified that all usages of this method are ok to ignore potential exceptions. Namely, as long as the test logic itself doesn't want to ensure that any topics got deleted, and as long as this method is the last line in the method, then it should be fine just to ignore failures here.

I also considered just deleting the method, but if it does succeed, then it leaves less garbage around for subsequent tests, so it feels better to at least attempt a cleanup.

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java [hunk 4]
```diff
@@ -106,7 +106,9 @@ public static void startKafkaStreamsAndWaitForRunningState(final KafkaStreams ka
         kafkaStreams.start();
         assertThat(
             "KafkaStreams did not transit to RUNNING state within " + timeoutMs + " milli seconds.",
-            countDownLatch.await(timeoutMs, TimeUnit.MILLISECONDS), equalTo(true));
+            countDownLatch.await(timeoutMs, TimeUnit.MILLISECONDS),
+            equalTo(true)
+        );
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java:111]: just fixing the formatting.

**Review by vvcephei — COMMENTED** [review]
Hey @guozhangwang , do you mind taking a look at this, since it overlaps with some of your recent contributions?

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:171]: req: Actually deleting topics after test is critical for some tests: I've encountered some cases where the same topics are reused mistakenly across different test cases within the single class. But I feel that it is better to put the topic deletion in the `@after` function while leaving `cleanUp()` as part of the test function itself.

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java [hunk 5]
```diff
@@ -106,7 +108,7 @@ public void before() {
 
         consumerConfiguration = new Properties();
         consumerConfiguration.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
-        consumerConfiguration.setProperty(ConsumerConfig.GROUP_ID_CONFIG, name.getMethodName() + "-consumer");
+        consumerConfiguration.setProperty(ConsumerConfig.GROUP_ID_CONFIG, safeTestName + "-consumer");
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]: Somewhere else it is set as `"group-" + safeTestName`, is this change intentional?

File: streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java [hunk 6]
```diff
@@ -215,13 +216,13 @@ public void onRestoreEnd(final TopicPartition topicPartition, final String store
...
-        final String applicationId = "streamsApp";
+        final String safeTestName = safeUniqueTestName(getClass(), testName);
         final Properties config = new Properties();
         config.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
-        config.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId + name.getMethodName());
+        config.put(StreamsConfig.APPLICATION_ID_CONFIG, "app-" + safeTestName);
         config.put(StreamsConfig.APPLICATION_SERVER_CONFIG, "localhost:" + (++port));
         config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, cluster.bootstrapServers());
-        config.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory(applicationId).getPath());
+        config.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java:225]: Is it safer to encode the appID as part of the dir path to avoid collision?

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java [hunk 7]
```diff
@@ -156,7 +157,7 @@ public void shouldPreservePartitionTimeOnKafkaStreamRestart() {
             assertThat(lastRecordedTimestamp, is(5000L));
         } finally {
             kafkaStreams.close();
-            cleanStateAfterTest(CLUSTER, kafkaStreams);
+            quietlyCleanStateAfterTest(CLUSTER, kafkaStreams);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java:160]: nit: we can put kafkaStreams in a try block.

File: streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java [hunk 8]
```diff
@@ -243,7 +251,7 @@ public void shouldRecoverBufferAfterShutdown() {
 
         } finally {
             driver.close();
-            cleanStateAfterTest(CLUSTER, driver);
+            quietlyCleanStateAfterTest(CLUSTER, driver);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java:254]: nit: ditto here, we can put `driver` in the try block. And ditto elsewhere.

**Review by guozhangwang — COMMENTED** [review]
Nice improvement. 

I made a pass; except the last `req` comment I do not feel strong about other nits, so feel free to ignore.

File: streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java (hunk 6 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java:225]: I wavered on this point, but each time you call `tempDirectory`, it should give you a completely new directory:
```
     * Create a temporary relative directory in the default temporary-file directory with the given prefix.
     *
     * @param prefix The prefix of the temporary directory, if null using "kafka-" as default prefix
```

So the prefix seems to be nice only for documenting which test a directory is for, not for enforcing any kind of test/directory uniqueness. I felt like it added more noise than value, so I just dropped all the prefixes.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:171]: I share your concern, but I'm not sure about the conclusion.

Yes, if there is state (such as a topic) that leaks from one test to the next, it can certainly cause difficult-to-debug failures. However, there are multiple things we can do to prevent/mitigate it:
* delete state after tests (not to leave any garbage behind)
* delete state before the tests (to ensure a clean slate for the test)
* choose unique names for all resources of each test (this is where the other part of this PR comes in)

Any one of these should be sufficient to prevent state from leaking in between tests, and most of these tests do all three. In other words, we have 3x redundancy guarding against such test pollution. If you look at all three of these measures, the clean up _after_ tests is actually the most optional, since tests can't tolerate failures in the clean up _before_ (because it also creates necessary topics), and choosing unique topic names per test is bulletproof and easy to fix (once we know what the problem is).

Whether the cleanup is part of the test or in the `@After` method, the outcome is the same, if the method throws an exception, the test will fail. The downside of After is that it requires you to store the topic names in mutable class-level fields, which actually makes it more awkward to choose unique names per test.

File: streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java:160]: Unfortunately, we generally can't use try-with-resources for these tests, since that makes the `kafkaStreams` reference out of scope for the finally block. We'd have to allow a reference to kafkaStreams to escape the try {} block to reference it either in finally or in an After method, which is just as messy as it currently is, if not more.

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java (hunk 5 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]:
```suggestion
        consumerConfiguration.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "group-" + safeTestName);
```

No, I just got tired of messing with every tests' idiosyncrasies.

**Comment by vvcephei** [comment]
Thanks for the review, @guozhangwang . I've addressed (or responded to) your comments.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:171]: Sounds good. I think this is convincing :)

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:171]: Whew! :)

File: streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java:225]: Yeah I was thinking about debugging purposes as well. If it is more noise the usefulness, I'm fine with dropping it.

**Comment by guozhangwang** [comment]
LGTM! Please feel free to merge after rebasing.

**Comment by vvcephei** [comment]
Thanks, @guozhangwang !
